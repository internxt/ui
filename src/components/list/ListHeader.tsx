import { ArrowDown, ArrowUp } from '@phosphor-icons/react';
import Checkbox from '../checkbox/Checkbox';
import { MenuType } from '../contextMenu/ContextMenu';

export type HeaderProps<T, F> = {
  label: string;
  width: string;
} & (
  | { name: F; orderable: true; defaultDirection: 'ASC' | 'DESC'; buttonDataCy?: string; textDataCy?: string }
  | { name: keyof T; orderable: false }
);

interface ListHeaderProps<T, F> {
  selectedItems: T[];
  items: T[];
  header: HeaderProps<T, F>[];
  isVerticalScrollbarVisible: boolean | null;
  orderBy?: { field: F; direction: 'ASC' | 'DESC' };
  menu?: MenuType<T>;
  displayMenuDiv?: boolean;
  checkboxDataCy?: string;
  onTopSelectionCheckboxClick: () => void;
  onOrderableColumnClicked: (column: HeaderProps<T, F>) => void;
}

const ListHeader = <T, F extends keyof T>({
  selectedItems,
  onTopSelectionCheckboxClick,
  items,
  header,
  orderBy,
  onOrderableColumnClicked,
  menu,
  displayMenuDiv,
  isVerticalScrollbarVisible,
  checkboxDataCy,
}: ListHeaderProps<T, F>) => {
  return (
    <div className="flex h-12 shrink-0 flex-row px-5">
      {/* COLUMN */}
      <div className="flex h-full min-w-full flex-row items-center border-b border-gray-10">
        {/* SELECTION CHECKBOX */}
        <div className="flex h-full flex-row items-center justify-between pr-4">
          <Checkbox
            checked={selectedItems.length > 0}
            indeterminate={items.length > selectedItems.length && selectedItems.length > 0}
            onClick={onTopSelectionCheckboxClick}
            checkboxDataCy={checkboxDataCy}
          />
        </div>

        {header.map((column) => (
          <button
            onClick={
              column.orderable
                ? () => {
                    onOrderableColumnClicked(column);
                  }
                : undefined
            }
            key={column.name.toString()}
            data-cy={'buttonDataCy' in column && column?.buttonDataCy}
            className={`flex h-full shrink-0  flex-row items-center space-x-1.5 text-base font-medium text-gray-60  ${
              column.width
            } ${column.orderable ? 'cursor-pointer hover:text-gray-80' : ''}`}
          >
            <span data-cy={'textDataCy' in column && column?.textDataCy}>{column.label}</span>
            {column.name === orderBy?.field &&
              column.orderable &&
              (orderBy?.direction === 'ASC' ? (
                <ArrowUp size={14} weight="bold" />
              ) : (
                <ArrowDown size={14} weight="bold" />
              ))}
          </button>
        ))}
        {isVerticalScrollbarVisible && <div className="mr-15px" />}
        {(menu || displayMenuDiv) && <div className="flex h-full w-12 shrink-0" />}
      </div>
    </div>
  );
};

export default ListHeader;
