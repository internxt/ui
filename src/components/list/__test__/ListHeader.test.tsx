import React, { render, fireEvent, screen } from '@testing-library/react';
import { describe, it, expect, vi, afterEach } from 'vitest';
import ListHeader from '../ListHeader';

const mockOnTopSelectionCheckboxClick = vi.fn();
const mockOnOrderableColumnClicked = vi.fn();

const renderListHeader = (props = {}) =>
  render(
    <ListHeader
      selectedItems={[]}
      items={[
        { id: 1, name: 'name1' },
        { id: 2, name: 'name2' },
      ]}
      header={[
        {
          label: 'Column 1',
          width: '100px',
          name: 'id',
          orderable: true,
          defaultDirection: 'ASC',
        },
        {
          label: 'Column 2',
          width: '200px',
          name: 'name',
          orderable: false,
        },
      ]}
      isVerticalScrollbarVisible={true}
      onTopSelectionCheckboxClick={mockOnTopSelectionCheckboxClick}
      onOrderableColumnClicked={mockOnOrderableColumnClicked}
      {...props}
    />,
  );

describe('ListHeader', () => {
  afterEach(() => {
    vi.clearAllMocks();
  });

  it('should match snapshot', () => {
    const listHeader = renderListHeader();
    expect(listHeader).toMatchSnapshot();
  });

  it('should render the columns correctly', () => {
    renderListHeader();

    expect(screen.getByText('Column 1')).toBeInTheDocument();
    expect(screen.getByText('Column 2')).toBeInTheDocument();
  });

  it('should toggle the checkbox state when clicked', () => {
    renderListHeader();

    const checkbox = screen.getByRole('checkbox');
    fireEvent.click(checkbox);

    expect(mockOnTopSelectionCheckboxClick).toHaveBeenCalledTimes(1);
  });

  it('should call onOrderableColumnClicked when clicking on an orderable column', () => {
    const { getByText } = renderListHeader();
    const column = getByText('Column 1');

    fireEvent.click(column);

    expect(mockOnOrderableColumnClicked).toHaveBeenCalledTimes(1);
    expect(mockOnOrderableColumnClicked).toHaveBeenCalledWith({
      label: 'Column 1',
      width: '100px',
      name: 'id',
      orderable: true,
      defaultDirection: 'ASC',
    });
  });

  it('should not call onOrderableColumnClicked for non-orderable columns', () => {
    const { getByText } = renderListHeader();
    const column = getByText('Column 2');

    fireEvent.click(column);

    expect(mockOnOrderableColumnClicked).not.toHaveBeenCalled();
  });
});
