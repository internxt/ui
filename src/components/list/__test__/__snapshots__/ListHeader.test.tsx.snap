// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ListHeader > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex h-12 shrink-0 flex-row px-5"
      >
        <div
          class="flex h-full min-w-full flex-row items-center border-b border-gray-10"
        >
          <div
            class="flex h-full flex-row items-center justify-between pr-4"
          >
            <label
              class="relative h-5 w-5 rounded focus-within:outline-primary undefined"
            >
              <div
                class="relative flex h-5 w-5 cursor-pointer flex-col items-center justify-center rounded border text-white border-gray-30 hover:border-gray-40"
              />
              <input
                class="base-checkbox h-0 w-0 appearance-none opacity-0"
                readonly=""
                type="checkbox"
              />
            </label>
          </div>
          <button
            class="flex h-full shrink-0  flex-row items-center space-x-1.5 text-base font-medium text-gray-60  100px cursor-pointer hover:text-gray-80"
            data-cy="false"
          >
            <span
              data-cy="false"
            >
              Column 1
            </span>
          </button>
          <button
            class="flex h-full shrink-0  flex-row items-center space-x-1.5 text-base font-medium text-gray-60  200px "
            data-cy="false"
          >
            <span
              data-cy="false"
            >
              Column 2
            </span>
          </button>
          <div
            class="mr-15px"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex h-12 shrink-0 flex-row px-5"
    >
      <div
        class="flex h-full min-w-full flex-row items-center border-b border-gray-10"
      >
        <div
          class="flex h-full flex-row items-center justify-between pr-4"
        >
          <label
            class="relative h-5 w-5 rounded focus-within:outline-primary undefined"
          >
            <div
              class="relative flex h-5 w-5 cursor-pointer flex-col items-center justify-center rounded border text-white border-gray-30 hover:border-gray-40"
            />
            <input
              class="base-checkbox h-0 w-0 appearance-none opacity-0"
              readonly=""
              type="checkbox"
            />
          </label>
        </div>
        <button
          class="flex h-full shrink-0  flex-row items-center space-x-1.5 text-base font-medium text-gray-60  100px cursor-pointer hover:text-gray-80"
          data-cy="false"
        >
          <span
            data-cy="false"
          >
            Column 1
          </span>
        </button>
        <button
          class="flex h-full shrink-0  flex-row items-center space-x-1.5 text-base font-medium text-gray-60  200px "
          data-cy="false"
        >
          <span
            data-cy="false"
          >
            Column 2
          </span>
        </button>
        <div
          class="mr-15px"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
